pac4j-sql/src/main/java/org/pac4j/sql/credentials/authenticator/DbAuthenticator.java

```diff
@@ -0,0 +1,173 @@
+ *
+ * @author Jerome Leleu
+ * @since 1.8.0
+ */
+public class DbAuthenticator extends AbstractUsernamePasswordAuthenticator {
+
+    protected final Logger logger = LoggerFactory.getLogger(getClass());
+
+    protected DBI dbi;
+
+    protected DataSource dataSource;
+
+    protected String attributes = "";
+
+    protected String startQuery = "select username, password";
+    protected String endQuery = " from users where username = :username";
+
+    public DbAuthenticator() {
+    }
+
+    public DbAuthenticator(final DataSource dataSource) {
+        this.dataSource = dataSource;
+    }
+
+    public DbAuthenticator(final DataSource dataSource, final String attributes) {
+        this.dataSource = dataSource;
+        this.attributes = attributes;
+    }
+
+    public DbAuthenticator(final DataSource dataSource, final String attributes, final PasswordEncoder passwordEncoder) {
+        this.dataSource = dataSource;
+        this.attributes = attributes;
+        this.passwordEncoder = passwordEncoder;
+    }
+
+    protected void initDbi() {
+        if (this.dbi == null) {
+            synchronized (this) {
+                if (this.dbi == null) {
+                    this.dbi = new DBI(this.dataSource);
+                }
+            }
+        }
+    }
+
+    public void validate(UsernamePasswordCredentials credentials) {
+        CommonHelper.assertNotNull("dataSource", this.dataSource);
+        CommonHelper.assertNotNull("attributes", this.attributes);
+        CommonHelper.assertNotNull("passwordEncoder", this.passwordEncoder);
+
+        initDbi();
+
+        Handle h = null;
+        try {
+            h = dbi.open();
+
+            final String username = credentials.getUsername();
+            final String query;
+            if (CommonHelper.isNotBlank(attributes)) {
+                query = startQuery + ", " + attributes + endQuery;
+            } else {
+                query = startQuery + endQuery;
+            }
+            final List<Map<String, Object>> results = h.createQuery(query).bind("username", username).list(2);
+
+            if (results == null || results.size() == 0) {
+                throw new AccountNotFoundException("No account found for: " + username);
+            } else if (results.size() > 1) {
+                throw new MultipleAccountsFoundException("Too many accounts found for: " + username);
+            } else {
+                final Map<String, Object> result = results.get(0);
+                final String expectedPassword = passwordEncoder.encode(credentials.getPassword());
+                final String returnedPassword = (String) result.get("password");
+                if (CommonHelper.areNotEquals(returnedPassword, expectedPassword)) {
+                    throw new BadCredentialsException("Bad credentials for: " + username);
+                } else {
+                    final DbProfile profile = createProfile(username, attributes.split(","), result);
```

Comment:
Your attributes in the sql query are basically just other extra fields. What you have here, I think, is  way to take those fields literally as they were defined, but that will not work if they are aliased. 

For example, what you have for attributes could be "firstname, lastname" and that would work. 

If I alias them in the query, which would "firstname AS F, lastname as L", that will not work when you do the parsing because you would get back F and L as names. So it's better to use the key for the attribute name off of the resulting map.